src/tools/rbd_mirror/ImageReplayer.cc

```diff
@@ -542,44 +553,34 @@ void ImageReplayer<I>::on_start_fail_finish(int r)
   Context *on_stop_finish(nullptr);
   {
     Mutex::Locker locker(m_lock);
-    if (m_stop_requested) {
-      assert(r == -EINTR);
-      dout(20) << "start interrupted" << dendl;
-      m_state = STATE_STOPPED;
-      m_stop_requested = false;
+    if (r < 0 && r != -EINTR) {
+      derr << "start failed: " << cpp_strerror(r) << dendl;
     } else {
-      assert(m_state == STATE_STARTING);
-      dout(20) << "start failed" << dendl;
-      m_state = (r < 0) ? STATE_UNINITIALIZED : STATE_STOPPED;
+      dout(20) << "start interrupted" << dendl;
     }
+
     std::swap(m_on_start_finish, on_start_finish);
     std::swap(m_on_stop_finish, on_stop_finish);
   }
 
-  update_mirror_image_status(true);
-
-  m_local_ioctx.close();
-  m_remote_ioctx.close();
-
-  if (on_start_finish != nullptr) {
-    dout(20) << "on start finish complete, r=" << r << dendl;
-    on_start_finish->complete(r);
-  }
-  if (on_stop_finish != nullptr) {
-    dout(20) << "on stop finish complete, r=" << r << dendl;
-    on_stop_finish->complete(0);
-  }
+  update_mirror_image_status(true, STATE_STOPPED);
+  handle_stop(r, on_start_finish, on_stop_finish);
 }
 
 template <typename I>
 bool ImageReplayer<I>::on_start_interrupted()
 {
-  Mutex::Locker locker(m_lock);
-  assert(m_state == STATE_STARTING);
-  if (m_on_stop_finish == nullptr) {
-    return false;
+  {
+    Mutex::Locker locker(m_lock);
+    assert(m_state == STATE_STARTING);
+    if (m_on_stop_finish == nullptr) {
+      return false;
+    }
   }
 
+  set_state_description(0, "");
```

Comment:
Do you really need this here? `on_start_fail_start(-EINTR)` below will call `set_state_description(-EINTR, "")`. The old code looks like still should work and is simpler due to avoiding the lock dance.